### ISSUE TITLE
Complete configuration being wiped

### ISSUE THREAD

**Issue by avsaase** [iNavFlight/inav#7128]
There seems to be an issue in 3.0 where the flight controller randomly loses its configuration. I had this problem once but assumed it was something I did wrong. However, since the 3.0 release, at least two other users on FB reported this problem so I think it is a bug.

The steps I took to get run into this problem were:

1. Have a `diff` from a previous 3.0 setup
2. Flash the 3.0 stable release
3. Choose "Airplane without a tail" platform
4. Restore `diff`
5. Power cycle with a battery (don't know if it happens with just USB power)
6. Notice the configuration has been wiped

The problem is I have done these exact steps 10's of times with the RC's and older beta builds, and only encountered this problem once. So far I have also not heard anyone running into this issue multiple times. I wouldn't know how to reproduce it again. 

I realize that this makes it very difficult to find the cause so I'm opening this issue to collect more experiences. Maybe it rings a bell for someone.

**Comment by semich** [comment]
Couple of point:
3 users complains about the same
all on F411-wse so far.
We setup the FC from scratch, all is ready for maiden.
Connect the FC again to inav for more checks of config and the FC is completely reset.
Matek says they are not aware of issues like this.
I had no issues on previous builds.

**Comment by avsaase** [comment]
I forgot to add that this happened on my Matek F405-WSE.

**Comment by OptimusTi** [comment]
Testing but quick note: I've been using a F411-WSE on 3.0 since 2.6 came out and have not seen the issue. Are we sure that the problem is not people forgetting to type save? :)

**Comment by avsaase** [comment]
I am sure that is not the problem.

**Comment by semich** [comment]
> Testing but quick note: I've been using a F411-WSE on 3.0 since 2.6 came out and have not seen the issue. Are we sure that the problem is not people forgetting to type save? :)

Seriously?! Sorry, we are very dumb peoples...

**Comment by Mateyhv** [comment]
Haven't seen this issue on F411-Wing... yet. Have reflashed it several times this week. The only difference from the above mentioned steps is that I do not do 3. Choose "Airplane without a tail" platform, but the rest is all the same.

**Comment by rts18** [comment]
Experienced it a few months back using development FW. Was an F411_wing. 

The Windows configurator has become very sluggish in 3.0. PID, Advanced and OSD tabs are very slow to open, And the reboots are slow sometimes too.

I've got into the habit of closing the configurator and power cycle the FC after a flash. And do the same after loading diff all files.

**Comment by b14ckyy** [comment]
> The Windows configurator has become very sluggish in 3.0. PID, Advanced and OSD tabs are very slow to open, And the reboots are slow sometimes too.

that's known because of a lot more stuff get's loaded on these pages compared to older versions. The developers know about this and will address it in the future. But that means a lot of rework how the Configurator communicates with the FC. For now we have to live with it. 

regarding the main issue: 
I have unfortunately (or luckily) not seen that yet on any of my planes. But Where I have seen a simular issue is on my U360GTS antenna tracker that is running on an SPRacing F3 witht he latest release candidate firmware. and it happened multiple times on one evening. Sometimes with battery power cycle, sometiomes with USB power cycle. All Config wiped.

So there is something that can indeed happen to the flash memory to lose all the config. The latest added feature to the tracker firmware is a home save option that can restore the last home point after a power loss. So the Tracker is writing to the flash memory while it is running. 

I wonder if there is something in INAV that can write to the flash even when no settings is changed by the user and the plane was not armed and corrupts the flash storage?

**Comment by OptimusTi** [comment]
It would be good to know the status of the LEDs     if    someone can reproduce the issue.

**Comment by avsaase** [comment]
I've run into the problem a couple of times when restoring (parts of) diffs, but I cannot reproduce it reliably.

**Comment by semich** [comment]
its random, maybe 1 in 10/20 reboots...

**Comment by OptimusTi** [comment]
> its random, maybe 1 in 10/20 reboots...

When you say "reboots" are you talking about rebooting through CMS or power cycles?

**Comment by Mateyhv** [comment]
>  restoring (parts of) diffs...

This is a different story, happened to me yesterday just after posting my previous message. Copied maybe 25 diff lines because I have seen some red warning message and wanted to see what was it about. Without saving, and just exiting the CLI and reconnecting to Configurator the setup message for selecting the type of vehicle appeared (as when starting from scratch).

**Comment by semich** [comment]
> > its random, maybe 1 in 10/20 reboots...
> 
> When you say "reboots" are you talking about rebooting through CMS or power cycles?

Yes. Unplug/ plus usb only, or battery only.

**Comment by OptimusTi** [comment]
> > > its random, maybe 1 in 10/20 reboots...
> > 
> > 
> > When you say "reboots" are you talking about rebooting through CMS or power cycles?
> 
> Yes. Unplug/ plus usb only, or battery only.

Cool. Next time it happens check the LEDs. I want to rule out a faulty DFU boot button.

**Comment by avsaase** [comment]
I think it it very unlikely that this is a hardware issue. While two people have told me they encountered this issue before 3.0, the number if posts about this since the last release

**Comment by Mateyhv** [comment]
It happened after a partial diff all load without saving and then reopen communications. Not even rebooting.
When full diff all and save things have always went smooth.

**Comment by OptimusTi** [comment]
I've noticed something strange. Following the steps requested to try and reproduce the issue, I noticed that after selecting the preset, I can go and click in the CLI or any tab. If I don't click away, the board reboots.  Could this cause issues or miss something if I click away before this reboot?

**Comment by rts18** [comment]
I ran some more tests today. This time with a F411WSE. I noticed a discrepancy in the reboot cycle after choosing these model defaults. Sometimes it reboots and other times it doesn't. 
Then after loading the diff all files and rebooting. It appears to complete the reboot for the model default. Which might explain why it clears the diff all settings.

As a side note. I have experienced the same issue with the OSD tab in 3.0. 
It used to do a disconnect and reconnect of the FC 7456 after loading the font prior to 3.0. This could be heard via the USB audio on the PC.  And the configurator would close down the OSD page on reboot, then reload the page again.
Now it disconnects after the font load, But the configurator doesn't always exit the OSD tab. If I try to change any of the font previews, it works for a bit then locks up until after a power cycle.

**Comment by NulliFieDFPV** [comment]
I am not sure if it's same issue.
Matek F405 WSE - Inav3.0.

Wipe didn't happen since flashed and several power cycles. 

But at todays maiden after a not that hard crash all my settings are gone, too. I thought it may happened because lipo and gps unit have been ripped off and ejected, but sounds a bit related to this.

Since I already set up everything again, I don't have a lot further details but will keep an eye on this. 

Regards, Ingo

**Comment by yesthisusernameistaken** [comment]
Suffering from this issue, these are the steps I take on a Matek 765. 
Connected over USB, not trying to load a diff, starting from a clean slate. 

Flash board with Inav using the latest configurator. 
Connect up and it asks me to choose a pre-set (wing without tail), FC reboots. 
Connect up again and change ports for GPS and RC link. Hit the save and reboot button. 

Sometimes the settings are saved, sometimes not. 
If the settings save, I got to the config tab to enable GPS. When I hit save, again sometimes it's saved, sometimes everything is reset to default (ports and other settings). 

I "think" the mixer gets reset to default too as it asks me to configure it after a couple reboots (I'm not sure if that little popup the first time you connect is supposed to do it or not). 

I've flashed it 2 or 3 times and the issues persist. 
Tried doing it with the battery plugged in and no difference.

**Comment by semich** [comment]
It just happened to me again. I flew the plane this morning, all well setup, all good and great! Arrived at home, plug the USB to check/modify one thing and the board is clean again.... (F411-wse)

**Comment by b14ckyy** [comment]
Another report. 

![PSX_20210621_143239.jpg](https://user-images.githubusercontent.com/33039058/122762389-8ea7eb80-d29d-11eb-8ddf-b8388cf711a6.jpg)

**Comment by avsaase** [comment]
Nope, still cannot reproduce. Neither with continuous trim on nor off.

**Comment by b14ckyy** [comment]
I tried on an F765 yesterday a lot. With presets applied, aborted by tab change, diff transferred and many other usual and unusual actions. Could not reproduce. Also tried a few times with an F411-WSE with no "luck". 

Will do more tests. F411WSE seems to be the most prominent one.

**Comment by P-I-Engineer** [comment]
just happened to me. Matek F411WSE.

the entire board was wiped. Did not even show up as a comport when plugged in to computer. I had to hold the boot button to get into DFU mode to reflash. Luckily, from knowing about this issue, i grabbed a diff yesterday.

last things i did:
i was in cli and changed some launch parameters, i think i forgot to hit save.

continuous servo autotrim is enabled.

**Comment by b14ckyy** [comment]
Now it hit me too.

Talon 250G. F411-WSE. No dump saved beforehand. It was flying fine last time. Today I wanted to connect it to the PC to get a diff for a friend and this screen welcomed me.

![Screenshot_20210702-185251.png](https://user-images.githubusercontent.com/33039058/124306465-b97c2480-db66-11eb-868b-fb628695d2e2.png)

Have to check the 2 other F411WSE setups too.

**Comment by OptimusTi** [comment]
Wondering of those new features that save to the auto eeprom are causing issues with this board

**Comment by P-I-Engineer** [comment]
when mine happened, it was the first time i'd ever had autolevel on.

**Comment by avsaase** [comment]
I don't think autolevel automatically saves to eeprom on disarm. Continuous servo trim does but iirc there was one report of this problem with that feature disabled. I could be wrong on that though.

**Comment by P-I-Engineer** [comment]
no clue, just adding info

**Comment by b14ckyy** [comment]
@avsaase do you know if CAT will cause always a write command or only if the values have changed?

**Comment by avsaase** [comment]
> @avsaase do you know if CAT will cause always a write command or only if the values have changed?

Always.

**Comment by mf0o** [comment]
Happend to me today too. F405-wse, stable INav 3.0. 
I was connected via USB to OSX, did reset the servo midpoint, saved via GUI button. 

After reconnecting everything was gone.

**Comment by avsaase** [comment]
It's pretty clear now that the problem occurs when the config is saved to EEPROM. Continuous trim saves to EEPROM on every disarm and since a lot of people are using this feature, this could potentially explain the frequency of reports.

I think we should look into `isEEPROMContentValid()` and the changes in made in https://github.com/iNavFlight/inav/pull/6384. This function runs on during initialization and if it returns `false`, the configuration is reset. @digitalentity?

**Comment by b14ckyy** [comment]
I talked to Mj Ke right now and Something came in my mind, especially after the recent Poll on FB. 

I see some kind of pattern. The most wiped FC is the F411-WSE. The second most (actually it feels like a factor of 20 less of a chance) are boards with F405 Processors. And I have not seen a single report of any F7 fully wiped yet. And this corellates to the CPU performance of the Processors (and likely also EEPROM read speed?).

based on what you found @avsaase if this sanity check resets the configuration if it finds a corruption, could this be some kind of timing error? With INAV 3.0 the Config size grew a lot compared to 2.6 and therefore the sanity check might take longer too. And if there is a timing issue someone the longer sanity check times on slower CPU might see the error and wipe with higher quantity than more modern and faster CPUs.

**Comment by willsteffen1911** [comment]
I just had this happen with my new F411wse..  Done is twice in two days, while I'm configuring.

**Comment by P-I-Engineer** [comment]
i noticed the configurator is a bit laggy on eeprom saving. Does the a stutter corrupt?

**Comment by b14ckyy** [comment]
> i noticed the configurator is a bit laggy on eeprom saving. Does the a stutter corrupt?

Mop the Configurator only sends a SAVE command to INAV. And then it waits for the reboot or the command to finish. The F411 CPUs are indeed slower. Hence the reboot till reconnect time is noticeably longer than on F405 or F7. This can lead to a stutter of the Configurator.

This was what brought me to the boot time idea on my comment above.

**Comment by b14ckyy** [comment]
Now we seem to have a case of full wipe on a H7 board too.
![PSX_20210715_083957.jpg](https://user-images.githubusercontent.com/33039058/125740972-20a2b0aa-c885-40d4-af51-bf916f0ef83f.jpg)

**Comment by harry1453** [comment]
There are now two reports that the board wouldn't reboot until put into DFU mode and reflashed. To me this is a red flag - could this potentially indicate a bigger flash corruption problem?

IIRC the STM32s don't have actual EEPROM, they use a region of their flash memory for persistent storage. Maybe something that is causing the config to become corrupt could also cause the *program* to become corrupt? Hence not booting until re-flashed

**Comment by avsaase** [comment]
That would also explain the other settings related problems. In increasing level of security we have random settings changing (sometimes to invalid values), all settings being lost and the firmware itself becoming corrupted. According to [this spreadsheet](https://docs.google.com/spreadsheets/d/1fTvYh1GtOG6mt7DxMy2l5pYqH7ewknWmdtmkg9b0IZU/edit#gid=0) and also my personal experience the first predates 3.0. I've heard a couple reports that the full wipe problem also occurred before 3.0 but obviously not as often as it does now. The firmware corruption is new to me and doesn't seem that common.

**Comment by MrD-RC** [comment]
I've just had this happen to me while making changes on my development/demo Matek F405-WING.

I was making changes in the programming tab, but clicking save did not seem to make any changes save. The GUI log:

```2021-07-20 @ 17:12:25 -- Running - OS: Windows, Chrome: 78.0.3904.87, Configurator: 3.0.1
2021-07-20 @ 17:12:25 -- Loaded release information from GitHub.
2021-07-20 @ 17:12:40 -- Serial port successfully opened with ID: 1
2021-07-20 @ 17:12:40 -- MultiWii API version received - 2.4.0
2021-07-20 @ 17:12:40 -- Flight controller info, identifier: INAV, version: 3.1.0
2021-07-20 @ 17:12:40 -- Running firmware released on: Jul 13 2021 18:40:37
2021-07-20 @ 17:12:40 -- Board: MF4S, version: 0
2021-07-20 @ 17:12:40 -- Unique device ID received - 0x39003b464b50192038314d
2021-07-20 @ 17:14:01 -- Unrecoverable failure of serial connection, disconnecting...
2021-07-20 @ 17:14:01 -- Failed to close serial port
2021-07-20 @ 17:14:07 -- Serial port successfully opened with ID: 2
2021-07-20 @ 17:14:13 -- MultiWii API version received - 2.4.0
2021-07-20 @ 17:14:13 -- Flight controller info, identifier: INAV, version: 3.1.0
2021-07-20 @ 17:14:13 -- Running firmware released on: Jul 13 2021 18:40:37
2021-07-20 @ 17:14:13 -- Board: MF4S, version: 0
2021-07-20 @ 17:14:13 -- Unique device ID received - 0x39003b464b50192038314d
2021-07-20 @ 17:14:13 -- PWM output is disabled. Motors and servos will not work. Use Outputs tab to enable!
2021-07-20 @ 17:14:13 -- Mixer not configured. Use Mixer tab to set it up!
2021-07-20 @ 17:18:27 -- EEPROM saved
2021-07-20 @ 17:18:27 -- Device - Rebooting
2021-07-20 @ 17:18:27 -- Serial port successfully closed
2021-07-20 @ 17:18:35 -- Serial port successfully opened with ID: 3
2021-07-20 @ 17:18:36 -- MultiWii API version received - 2.4.0
2021-07-20 @ 17:18:36 -- Flight controller info, identifier: INAV, version: 3.1.0
2021-07-20 @ 17:18:36 -- Running firmware released on: Jul 13 2021 18:40:37
2021-07-20 @ 17:18:36 -- Board: MF4S, version: 0
2021-07-20 @ 17:18:36 -- Unique device ID received - 0x39003b464b50192038314d
2021-07-20 @ 17:18:36 -- PWM output is disabled. Motors and servos will not work. Use Outputs tab to enable!
2021-07-20 @ 17:18:36 -- Mixer not configured. Use Mixer tab to set it up!
```

This could be interesting? **2021-07-20 @ 17:14:01 -- Unrecoverable failure of serial connection, disconnecting...**

I couldn't see any other logs in the folders.

**Comment by SuperKohl** [comment]
Hey guys, today it hit me too, inav 3.0 and f411-wse

**Comment by harry1453** [comment]
Hey, just had this happen at the field. Only change was setting motor idle power, resulting in the following:

```
# status
System Uptime: 29 seconds
Current Time: 2041-06-28T02:04:00.000+01:00
Voltage: 0.00V (1S battery - NOT PRESENT)
CPU Clock=96MHz, GYRO=MPU6000, ACC=MPU6000, BARO=BMP280
STM32 system clocks:
  SYSCLK = 96 MHz
  HCLK   = 96 MHz
  PCLK1  = 48 MHz
  PCLK2  = 96 MHz
Sensor status: GYRO=OK, ACC=OK, MAG=NONE, BARO=OK, RANGEFINDER=NONE, OPFLOW=NONE, GPS=UNAVAILABLE
Stack size: 6144, Stack address: 0x20020000, Heap available: 1728
I2C Errors: 0, config size: 7193, max available config: 16384
ADC channel usage:
   BATTERY : configured = ADC 1, used = ADC 1
      RSSI : configured = ADC 3, used = none
   CURRENT : configured = ADC 2, used = ADC 2
  AIRSPEED : configured = ADC 4, used = none
System load: 6, cycle time: 1009, PID rate: 991, RX rate: 49, System rate: 9
Arming disabled flags: NAV HWFAIL RX CLI SETTINGFAIL
### ERROR: Invalid setting: gyro_hardware_lpf
VTX: band: A, chan: 2, power: 1 (25 mW), freq: 5845 MHz
```

**Comment by harry1453** [comment]
Hit me again today after saving a waypoint mission to EEPROM.

**Comment by Cris-Dev-I-Not** [comment]
Hello Everyone, 
Thanks to the devs for all the great work on inav. I appreciate you. Has anyone found a workaround for this yet? Also, is there a target release date for the fix? I am able to replicate this by just attempting to use SA. BUT I have had times where the config is lost sitting on the bench. This has happened to me in 2.6.1 as well? OR aM i cRazY peEEEEp. I have tried removing GPS, VTX, RX, and cam one by one with same results. 
Hardware profile:
ViFly Finder 2 Buzzer
Foxeer Razer Mini 1200TVL
RDQ Mach 3 Video Transmitter 25-1000mW 5.8GHz VTX
RDQ BN-880 Flight Control GPS Module w/ Compass	
Matek F411-WSE
AMAX F40A 40A 2-4S Wing ESC

**Comment by P-I-Engineer** [comment]
another config wipe for me yesterday on my bench FC.

iflight succex f4 - betaflightF4 target - F405 MCU

only basic config on fc- ports, gps, rx protocol, programming logic

**Comment by breadoven** [comment]
Outputs tab in Configurator breaks every time I flash now. Only fix is setting to Standard ESC protocol in the CLI then setting DShot in the Outputs tab. Setting DShot in the CLI after setting Standard, Saving and rebooting doesn't fix the Outputs tab though, you have to set DShot in the Outputs tab to get it working with DShot. Odd.

**Comment by jaroszmm** [comment]
Another 411wse clean wipe here. In my case happened after a couple (5-6) quick lipo on / off connections (was trying sth out) and eventually osd started showin gyro not calibrated so a complete wipe occured.

**Comment by MrD-RC** [comment]
Just happened to me again using the current master. I was again in the programming tab, changing settings in the PID controllers. To make the settings stick you have to power cycle for some reason. Anyway, power cycled and the defaults dialog appeared.

MATEK F405-WING (MATEKF405SE)
Just using USB power
Version: Master on this branch https://github.com/iNavFlight/inav/tree/MrD_Add-precision-to-mAh-used-for-9999mAh

**Comment by DzikuVx** [comment]
I had this today and I think I know why it's happening: power cycle (intentional or not) during save to EEPROM.In this case it was caused by a USB-C cable that was not fully inserted into the socket. I bumped the board while saving and it was enough

**Comment by harry1453** [comment]
@DzikuVx how long does writing to EEPROM take? Would it be possible to speed it up so that there is less of a likelihood of this happening?

**Comment by stronnag** [comment]
IIRC, from issuing the MSP command to the OS noticing the device node has been torn down (i.e. reboot in progress), around 1.3 seconds. I'll redo the experiment later today and confirm if no other data appears,

**Comment by stronnag** [comment]
It's also worth noting that in the last 18 months or so we have greatly increased the volume of data saved (more OSD elements, introduction of logic conditions / GVARs, increased no of waypoints). Aided and abetted by the fact that we also dropped architectures with less than 8KB of EEPROM space; supported targets are now at least 16K.

**Comment by stronnag** [comment]
> @DzikuVx how long does writing to EEPROM take? Would it be possible to speed it up so that there is less of a likelihood of this happening?

For a set-save-reboot loop test, in an attempt to find a way to reproduce the issue (which I could not).

```
Add device: /dev/ttyACM0 (INAV, FURIOUS_F35-LIGHTNING)
Read value = 0 (0)
Saving ...
Save took 1.285923437s
Remove device: /dev/ttyACM0
Add device: /dev/ttyACM0 (INAV, FURIOUS_F35-LIGHTNING)
Read value = 1 (1)
Saving ...
Save took 1.295825802s
...
```

So yes, c. 1.3 seconds from issuing `EEPROM_WRITE` to the request returning. Let me similarly instrument the CLI version.

**Comment by stronnag** [comment]
Similar magnitude via CLI (from issuing `save`) until the FC replies `Rebooting`.
```
Add device: /dev/ttyACM0 (INAV, FURIOUS_F35-LIGHTNING)
Read value = 0 (0)
Save took 1.207931007s
Remove device: /dev/ttyACM0
Add device: /dev/ttyACM0 (INAV, FURIOUS_F35-LIGHTNING)
Read value = 1 (1)
Save took 1.214872209s
Remove device: /dev/ttyACM0
```
F405 in both cases.

**Comment by DzikuVx** [comment]
That basically means, if power is interrupted in those 1.2s, EEPROM will be corrupted and defaults restored on next boot

**Comment by b14ckyy** [comment]
I just tried it with a spare F405 Wing. When I enter save in the CLI and unplug USB in under 1.2s (before rebooting pops up in CLI) I can indeed corrupt the config and the initial setup appears. All wiped. 

But I have to be really fast. How can this happen during 2 flights? When I disarm I have my hands on the radio and I find it nearly impossible to unplug the battery within even 2s after disarm. Or does the disarm process maybe take longer? 

save time, disarm delay at default 250ms, is 1.5s in total. Is there more going on that can delay the save process further?

**Comment by Jetrell** [comment]
I've hardly experienced this issue at all until today.
I loaded a 5.0 master build from _May 14_ onto a quad with an Omnibus F4Pro. Using the 5.0, _May 14_ Configurator build.

After changing settings a few times, it seemed okay. Until I went into the CLI to check the values of some settings. Then typed exit.
After a power up. It was back to default. 
This happened twice on two separate occasions. 

This board and its USB plug is in perfect condition.. Never crashed, and no grass/dirt or damage.
And the USB lead I used today was also used on two other flight controllers, that did not, and have never experienced settings being reset to default.

**Comment by stronnag** [comment]
The number of logic conditions was increased recently (May 10th); however the corresponding `PG_VERSION` was not incremented (just _too easy_ to forget). Maybe this is related.

**Comment by MrD-RC** [comment]
Fixed

**Comment by DzikuVx** [comment]
Thanks Darren

**Comment by dcamptell** [comment]
im running inav 4 and 5 and this has happened to me multiple times and sometimes within 3 or 4 power cycles its getting very frustrating. i have a dump file to restore, but being out in the feild ready to fly and its all gone not very fun i hope this gets fixed by the release of inav 5 thanks to all  the devs for all their hard work!

**Comment by b14ckyy** [comment]
@dcamptell you have to make sure you give your FC time to save the config before you power it down. Especially on F411 Boards the save process and checksum creation can take up to 3s. Allt he recent config wipes I have seen that happened for others where caused by powering down the FC too quickly after clicking safe or in one case after a crash. The Crash case happened because the Pilot disarmed short before impact (with Continuous auto trim active, so that disarm triggers a safe) and then the crash ejected the battery and this wiped the config. 

There is not much you can do about it beside just wait a bit before you power down the FC after making any change or after disarming.

**Comment by ahmadhassan11** [comment]
hi 
The value of p_roll in pid settings automatically becomes zero after refresh
Someone can guide me to fix this problem
fixed wing in inav 4

**Comment by Mateyhv** [comment]
@ahmadhassan11 you aren't providing much information about your case, and your issue doesn't fit very well into "complete configuration wiped out". P_roll becomes zero after refreshing what? What board are you using? If you are working with the configurator make sure you hit the save button before changing tabs.

**Comment by dcamptell** [comment]
My config was complete and had been powered up a few times without being
plugged in via USB started it up and tested everything and it was fine.
Took it to the feild and nothing when I plugged battery in. I'll see if it
happens again and watch for the continue trim on the servo on a crash
thanks for the reply!

**Comment by 0crap** [comment]
Just for the record, this happened to me and my buddy as well.
I'm using a Matek H743-wing v2, iNav 5.0
Plug in the USB cable to check a setting, bang, wiped config.
Got the screen where I had to select "Airplane without a tail".
Seen this only once, unable to reproduce the issue.

My buddy plugged in a fresh lipo after a successful flight, bang, wiped config.
Needed to past in a diff all to get it working again.
He uses a Omnibus flight controller and still on iNav 4.x
Also happened once, unable to reproduce the issue.

We also fly a lot of quads with Betaflight.
Never got a config wiped with BF.
(Just giving as much info as we have.)

**Comment by cpwilson** [comment]
Same issue here now, never seen it before on either my 411 or 405 boards on 4.1, but since upgrading the 405 to 5.0 it looks to be happening all the time.
Config saved and unplugged usb after a while to ensure save is completed.    Power up board from flight battery and all is OK initially but it seems that if left overnight it's defaulted when powered up.    Happened 3 times now so reverted back to 4.1 and no issues at all.

**Comment by breadoven** [comment]
Looks like I missed the config version bump for https://github.com/iNavFlight/inav/pull/7270. Is this likely to cause this ?

Also for https://github.com/iNavFlight/inav/pull/6850 ?

**Comment by Mateyhv** [comment]
@cpwilson to track down possible cuases would be usefull to add at least some additional info: exact board model, operating system, configurator version, type of USB on your computer (v2, v3) and for consistency type of USB connector on the board, and finally a 'diff all' file of the configuration been wiped out. Thanks

**Comment by cpwilson** [comment]
@Mateyhv 

Board is a Matek F405WSE - http://www.mateksys.com/?portfolio=f405-wse
OS being used is Windows 11
Configurator V5.0.0 64bit windows version
USB is Intel USB 3.0
Cable is USBC to the FC

Diff All attached
[INAV_cli_CW_NANO TALON_20220707_215536.txt](https://github.com/iNavFlight/inav/files/9076414/INAV_cli_CW_NANO.TALON_20220707_215536.txt)
d

**Comment by 0crap** [comment]
With iNav 5.0 we also have [this PR](https://github.com/iNavFlight/inav/pull/7624) that forces a lot of extra write actions.
I don't see the benefit of that PR any way, only doing a lot of unnecessary writes that might trigger the config wipe issue.
Not a fan of that one.

**Comment by DzikuVx** [comment]
@0crap this is disabled by default and you have to enable this function to have those writes. So I'm not sure why so negative attitude

**Comment by b14ckyy** [comment]
@DzikuVx the function is off (actually the init_gyro_cal=ON) by default but the gyro calibration values are always saved nevertheless. Are they saved when the calibration is completed or on disarm? 

If the calibration is stored after it is completed this could be a possible reason for a wipe if the user powers up the FC via USB but then powers it off right after the calibration is done. 
![image](https://user-images.githubusercontent.com/33039058/178112674-1ae95a6b-e7c9-4eb7-93c5-008edc23526b.png)

**Comment by b14ckyy** [comment]
oh shoot. yes I can confirm and reproduce. When I connect my F411-WSE via USB, count to approx 5s and then unplug, I get a full wipe. Just saced a diff to test it. 
![image](https://user-images.githubusercontent.com/33039058/178112787-0b73fea0-1c40-4555-afcf-70f6a2e190d1.png)

**Comment by digitalentity** [comment]
The thing is that we can't write the config without erasing is completely first. We need to harden our configuration writes. I see two options:
1. Keep two copies of the config and use the first valid one. Erase/write first copy, verify that it's valid then (and only then) erase/write the second one. 
2. Append PG copies to the end of the config (when possible). We read the whole config anyway and if it contains two copies of the same PG the last will take precedence.

A combination of both approaches would be even better.

**Comment by b14ckyy** [comment]
@digitalentity 
I like Option 1! 
But in general I think that this feature is now ALWAYS write a config any time the FC is started is a big problem in itself. Over time this will also degenerate the flash for 99% of people who don't use the stored value as they are saved over and over again. 

IMHO it should only store one single time if the stored calibration values are 0 and only if the startup calibration is turned OFF.

**Comment by 0crap** [comment]
> @DzikuVx the function is off (actually the init_gyro_cal=ON) by default but the gyro calibration values are always saved nevertheless. Are they saved when the calibration is completed or on disarm?
> 
^ Exactly this.
Even if you do not use that function the next `diff all` you pull is always different for:

```
set gyro_zero_x = -5
set gyro_zero_y = 22
set gyro_zero_z = -25
set ins_gravity_cmss =  957.532
```
Why?

**Comment by JulioCesarMatias** [comment]
> > @DzikuVx the function is off (actually the init_gyro_cal=ON) by default but the gyro calibration values are always saved nevertheless. Are they saved when the calibration is completed or on disarm?
> 
> ^ Exactly this. Even if you do not use that function the next `diff all` you pull is always different for:
> 
> ```
> set gyro_zero_x = -5
> set gyro_zero_y = 22
> set gyro_zero_z = -25
> set ins_gravity_cmss =  957.532
> ```
> 
> Why?

You can download the iNav algorithm, open it in VSCode and remove out the Gyro calibration save part that I added, compile and test! If the problem you're having goes away, ok, no problem, I'll try to fix my PR!

**Comment by JulioCesarMatias** [comment]
Don't just speculate where the problem might be without being sure, JUST compile the algorithm and test it!

**Comment by b14ckyy** [comment]
@0crap I took that topic just in the live stream. He said he will look into it and there will be a change. His idea is to not auto save the values but make them user save manually in CLI or with stick command instead. 
@JulioCesarMatias thanks for quickly looking into it. I think if whoever wants to use the feature just one time after runtime calibration uses the stick command for save or enters save in CLI, the values are stored with it as well. So the automatic save command could be just removed I guess?

**Comment by 0crap** [comment]
Awesome.
I can't reproduce the config wipe issue.
It's random here.

But in general I think it's a good idea not to write if it's not needed.
Especially if it MIGHT be related to the config wipe in some way.
Great if the new gyro calib. function can be improved.

**Comment by stronnag** [comment]
"Unsolicited` EEPROM write considered harmful" ; particularly as they take c. 1.2s to complete.

**Comment by b14ckyy** [comment]
I was able to reproduce it with a 50% error rate now by just counting to 5 between power on and power off in my head. 1.2s is a big time frame to hit if the FC is powered on and off again and this can happen very often if someone is for example setting the FC up.

**Comment by digitalentity** [comment]
I'm inclined to disable all "unsolicited" writes for all boards without a separate memory chip (not Blackbox, but a separate 24C16 or similar). Even if the configuration is invalid, we should initialize the runtime config from the defaults, but do not touch the flash memory until a "SAVE" button is pressed.

**Comment by b14ckyy** [comment]
@digitalentity maybe disable the autotrim automatic save as well? But the STATS save command on disarm needs to stay still.

**Comment by stronnag** [comment]
> @digitalentity maybe disable the autotrim automatic save as well? But the STATS save command on disarm needs to stay still.

Why? I'm not interested in 'stats'. I'd prefer to be in control of the wear of my EEPROM.

**Comment by b14ckyy** [comment]
@stronnag stats are disabled by default anyway. But if I want them for example so in this case if I enable them it's my liability as a pilot if I wear down my eeprom. If the autosave of continuous servo trim is disabled, this would reduce the automatic config writes down to 0 for everyone who does not explicitly enable stats.

**Comment by 0crap** [comment]
> > @digitalentity maybe disable the autotrim automatic save as well? But the STATS save command on disarm needs to stay still.
> 
> Why? I'm not interested in 'stats'. I'd prefer to be in control of the wear of my EEPROM.

Same here.
Stats I can pull from my DVR or Taranis logs.

**Comment by b14ckyy** [comment]
@0crap I like to have a record of the overall flight time and flown distance for my planes where I don't have to write it down by hand every time I fly it ad add it together by hand. Stats are enabled on all my planes. 
As I said they are off by default so I see no problem. If I'd argue like you I could also say: "Remove the Crossfire support because I don't need it anyway" ;)

**Comment by 0crap** [comment]
Continuous servo trim is the best best thing since sliced bread.
But if I'm pressed to make choice, EEPROM reliability comes first.

**Comment by stronnag** [comment]
>  stats are disabled by default anyway. 

If it's disabled by default, that's fine; it's not unsolicited, the pilot has elected to take the risk.

**Comment by 0crap** [comment]
> @0crap I like to have a record of the overall flight time and flown distance for my planes where I don't have to write it down by hand every time I fly it ad add it together by hand. Stats are enabled on all my planes. As I said they are off by default so I see no problem. If I'd argue like you I could also say: "Remove the Crossfire support because I don't need it anyway" ;)

Agreed.

**Comment by b14ckyy** [comment]
Yeah continuous servo trim can stay. Not saving the new trim after each flight does not reduce it's functionality at all. So just the auto save should be removed but the feature has to stay of course. After the maiden you save the trims once with stick command and these will be the starting trims anyway.

**Comment by 0crap** [comment]
> Yeah continuous servo trim can stay. Not saving the new trim after each flight does not reduce it's functionality at all. So just the auto save should be removed but the feature has to stay of course. After the maiden you save the trims once with stick command and these will be the starting trims anyway.

That would be a nice trade off, if you want them to stick, do a save and reboot when landed before you disconnect the battery.

**Comment by b14ckyy** [comment]
So as a summary we would agree  on: 
- disable the automatic save of gyro cal data (@JulioCesarMatias already sent me a changed build where it only saves one time and if the values are not 0 it won't save again unless the pilot does a save command by himself)
- disable the autosave from continuous servo trim (has to be saved manually after maiden to have a good starting point) 
- disable autosave of the regular servo Autotrim(??) 
- keep stats as it is because they are disabled by default

this should completely eliminate the chance of an accidental config wipe unless the pilot has stats an and disconnects battery right after disarm. Or save and reboots and disconnects from PC. 

Can't think of anything else that is automatically saved currently.

**Comment by 0crap** [comment]
LGTM!

**Comment by dcamptell** [comment]
I saved my config and waited 5 minutes and even clicked the
disconnect button, and waited another 30 seconds after I closed the
configurator and unplugged the fc. Still happened after 4 or 5 power
cycles.

**Comment by 0crap** [comment]
Maybe it's an idea to let the blue LED blink a different pattern indicating EEPROM saving is active.
Probably nobody will notice it, but for testers it takes out the guesswork when a save is actually happening.

**Comment by dcamptell** [comment]
I'm 100 percent sure 5 minutes is plenty of time to save I tried waiting Nd
just today wiped right when I tool it out to fly it. I did multiple tests
at home and it worked but the second I take it to the field it's gone. It's
completely random.

**Comment by b14ckyy** [comment]
@dcamptell you might consider a defect FC. 
This is just too much in your case. Many of us use 5.0 now and the wipe is still very rare in total. If you have a FC that wipes itself like every 5th boot there must be something messed up on the FC or in your setup.

**Comment by cpwilson** [comment]
@b14ckyy you suggest it may be FC related or something in the setup, but for me the FC is over 12 months old with no prior issues, have imported a diff all from 4.1 as recommended and problems until 5.0 went on and conig gets wiped all the time.
Rolling back to 4.1 makes the config wipe go away.   I can't see how it's hardware from my experience.

There must be something a miss in 5.0 as I seem to have the same results as @dcamptell with my F405wse
Mine is fine on the bench the night before but when I arrive at the field the day after the config is wiped, which has happened 3 times now.

@dcamptell have you tried rolling back to your prior version?

Strangely I can't replicate it at home with power cycles but having gone back to 4.1 yesterday all is well and had a few flights this morning with no issues at all so nothing wrong with my hardware.

**Comment by 0crap** [comment]
@dcamptell @cpwilson can you give the output of the `version` command in the CLI?
Maybe @JulioCesarMatias can compile you a 5.0 test version without the calibration savings. 
If that reduces the wipe issue for your setups, it might help the devs narrow it down.

**Comment by cpwilson** [comment]
# version
# INAV/MATEKF405SE 5.0.0 Jun 20 2022 / 11:00:33 (d2adf35c)
# GCC-10.2.1 20201103 (release)

**Comment by dcamptell** [comment]
This has happened to multiple fc's that I have just happened to my matek
f405 wte.

**Comment by 0crap** [comment]
> I was able to reproduce it with a 50% error rate now by just counting to 5 between power on and power off in my head. 1.2s is a big time frame to hit if the FC is powered on and off again and this can happen very often if someone is for example setting the FC up.

I tested it on the bench and found a way to reproduce it almost always. (Old Matek F405-STD board.)
Connect the FC to the iNav Configurator, find a tab where you can click on the Save button, directly after that pull the USB cable out.
Next plugin you will get the "Default values" screen before you can continue.

**Comment by MrD-RC** [comment]
@0crap Doing that will always lead to a config wipe. No matter what. If you don't allow the data to finish saving, it has no chance. That's not really a fair test or the bug. That's like testing if a tyre valve is leaking by stabbing the tyre.

**Comment by OptimusTi** [comment]
It's there a list of things that "auto-save"?

**Comment by 0crap** [comment]
> @0crap Doing that will always lead to a config wipe. No matter what. If you don't allow the data to finish saving, it has no chance. That's not really a fair test or the bug. That's like testing if a tyre valve is leaking by stabbing the tyre.

True.
But it might point into the direction of all the auto-saving stuff.
End users have no clue when an autosave is done and pull cables at any time.

**Comment by breadoven** [comment]
How significant is it if the PG_REGISTER_WITH_RESET_TEMPLATE version isn't changed ? I've noticed there are a few changes where this wasn't done including one of mine (https://github.com/iNavFlight/inav/pull/6850) also https://github.com/iNavFlight/inav/pull/8027. I thought this was only an issue of you don't do a full chip erase ? I assume you only need a PG_Version increment per INAV version release ?

And does this only appear to be a Matek board related issue or is it simply that so many people use Matek boards ? I did see that @Jetrell had it happen on a Omnibus board but that seems the only non Matek example. Never happened to me but I don't use any Matek boards.

**Comment by Jetrell** [comment]
Concerning the Omnibus F4 Pro this occurred with. I actually found what setting was causing the wipe.. It was the `motor_pwm_rate` change. 
This would occur with multiple computers.

Once I slowed it down to Oneshot125, with its default `pwm_rate`.. It hasn't happened again.
And to be sure, I went back and tried earlier firmware as before, and the issue returned.

**Comment by eabrega** [comment]
I have this problem too. Matek 411WSE, INAV 5.0. Settings has ben lost after connect power.

**Comment by MartinHugh** [comment]
I experienced total config wipe once on 4.1.0 with a Matek target MATEKF405SE

I had a hard unscheduled landing which caused the battery to eject.

Does that fit the hypothesis of power loss during EEPROM write?

I did have continuous autotrim on, but was in manual mode, I understood that the trim value is only saved when the flight mode is changed.

Any ideas what the most likely cause of this config loss was ?

**Comment by b14ckyy** [comment]
@MartinHugh the config is saved every time you  disarm if CAT is on. this might be also changed as it is actually not necessary to save the trims every time. you can save them once when autotuned as a baseline and then this is just fine adjustment.
but yes if you crash and disarm right before the impact, this triggers a save and corrupts the cinfig if the battery ejects on impact

**Comment by 0crap** [comment]
I commented out line 550 in src/main/flight/servos.c and compiled my own hex.
I'm not a coder but that should do it.

Will test in flight if it works as expected.

**Comment by MartinHugh** [comment]
> @MartinHugh the config is saved every time you disarm if CAT is on. this might be also changed as it is actually not necessary to save the trims every time. you can save them once when autotuned as a baseline and then this is just fine adjustment. but yes if you crash and disarm right before the impact, this triggers a save and corrupts the cinfig if the battery ejects on impact

Thank you for your thoughts, There is an excellent chance I did *not* disarm before impact.  Probably soon after, but too busy trying to save an aircraft close to the ground to get to the arming switch before the impact.

**Comment by MartinHugh** [comment]
> I commented out line [552](https://github.com/iNavFlight/inav/blob/master/src/main/flight/servos.c#L552) in [src/main/flight/servos.c](https://github.com/iNavFlight/inav/blob/master/src/main/flight/servos.c) and compiled my own hex. I'm not a coder but that should do it.
> 
> Will test in flight if it works as expected. (Looked at PR [6708](https://github.com/iNavFlight/inav/pull/6708) to figure out how it's done.)

Sorry that this is off thread, but is there a guide for setting up the tool chain to compile iNav.  Something I would like to do.

**Comment by 0crap** [comment]
> > I commented out line [552](https://github.com/iNavFlight/inav/blob/master/src/main/flight/servos.c#L552) in [src/main/flight/servos.c](https://github.com/iNavFlight/inav/blob/master/src/main/flight/servos.c) and compiled my own hex. I'm not a coder but that should do it.
> > Will test in flight if it works as expected. (Looked at PR [6708](https://github.com/iNavFlight/inav/pull/6708) to figure out how it's done.)
> 
> Sorry that this is off thread, but is there a guide for setting up the tool chain to compile iNav. Something I would like to do.

Have a look here: https://github.com/iNavFlight/inav/tree/master/docs/development
Currently I have it running in [Docker](https://github.com/iNavFlight/inav/blob/master/docs/development/Building%20in%20Docker.md) on my Windows box.
It's not easy to setup if you have never done it.
I also used the Betaflight documentation on GitHub to get it running.
They have a bit more documentation to get it going. Good luck!

**Comment by JulioCesarMatias** [comment]
@0crap I made some changes to iNav EEPROM, could you test and tell me if the problem still persists? Tell me what target are you using so I can compile

**Comment by 0crap** [comment]
> @0crap I made some changes to iNav EEPROM, could you test and tell me if the problem still persists? Tell me what target are you using so I can compile

Sure, but the problem is I rarely run into the config wipe issue randomly.
Can you explain what changes you made?

In the meanwhile my buddy and I tested a [build](https://github.com/0crap/inav/releases/tag/5.1.0) I created.
It's a build with all that is currently sheduled for iNav 5.1.0 but on top of that the continuous servo autotrim EEPROM saving is removed.
We tested that by doing a flight with continuous servo autotrim enabled, land and not disarm yet.
Pickup the plane and flick the disarm switch on the radio, then quickly pull the battery.
No issues, seems to work really well!
Continuous servo autotrim still works in flight. Awesome!

**Comment by JulioCesarMatias** [comment]
> Sure, but the problem is I rarely run into the config wipe issue randomly. Can you explain what changes you made?
> 
> In the meanwhile my buddy and I tested a [build](https://github.com/0crap/inav/releases/tag/5.1.0) I created. It's a build with all that is currently scheduled for iNav 5.1.0 but on top of that the continuous servo autotrim EEPROM saving is removed. We tested that by doing a flight with continuous servo autotrim enabled, land and not disarm yet. Pickup the plane and flick the disarm switch on the radio, then quickly pull the battery. No issues, seems to work really well! Continuous servo autotrim still works in flight. Awesome! Target is MATEKF405.

My changes checksums each PG section when it's read from EEPROM (FLASH) and a check is made in `writeSettingsToEEPROM()` to only perform a write if the checksum of any PG section has changed. 
[inav_6.0.0_MATEKF405.zip](https://github.com/iNavFlight/inav/files/9127617/inav_6.0.0_MATEKF405.zip)

**Comment by 0crap** [comment]
That sounds like a clever idea.
Does that mitigate the wipe issue if you pull the battery at the wrong moment when it writes to EEPROM?
The problem with continuous servo autotrim is that it always changes the servos a few micro seconds.
So with your change it still thinks a write is needed I guess.
To avoid continuous servo autotrim to write is still the best bet I believe.
(It also makes no sense to write for example 1503, 1502,1499.... while 1500 is already in your config.)
Or do I miss something?

**Comment by JulioCesarMatias** [comment]
> That sounds like a clever idea. Does that mitigate the wipe issue if you pull the battery at the wrong moment when it writes to EEPROM? The problem with continuous servo autotrim is that it always changes the servos a few micro seconds. So with your change it still thinks a write is needed I guess. To avoid continuous servo autotrim to write is still the best bet I believe. (It also makes no sense to write for example 1503, 1502,1499.... while 1500 is already in your config.) Or do I miss something?

This can alleviate the problem, yes, since the recording time will be faster... Because the recording will only occur in the sections that had the values ​​changed...

**Comment by 0crap** [comment]
> 
> This can alleviate the problem, yes, since the recording time will be faster... Because the recording will only occur in the sections that had the values ​​changed...
> 
Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all.
Definitely recommended for v5.1.0 if you ask me.
>
> About the close values ​​you mentioned, maybe we can add a "dead band" to this.
I would not opt for that, it removes some of the accuracy of he servo auto trim feature.
It's just I don't see the need to save it to EEPROM every flight.
If removing the autotrim save is out of the question, I would just go with your checksum solution and call it a day. 💯

**Comment by JulioCesarMatias** [comment]
> Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.

but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.

**Comment by P-I-Engineer** [comment]
Is continuous Auto trim only used for flying in manual? I don't have manual mode anymore, so continuous Auto trim prob isn't doin anything for me. Other than confirming my servo centerness.

**Comment by Mateyhv** [comment]
Auto trim is for Acro

**Comment by 0crap** [comment]
> > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> 
> but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.

I understand.
Problem is that I seldom run into the config wipe issue.
And the last time it was because we unplugged a lipo just after disarm.
So I'm not able to test and give a conclusive answer. Sorry.

A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates.
They run a MATEKF405SE and MATEKF405TE_SD

**Comment by 0crap** [comment]
> Is continuous Auto trim only used for flying in manual? I don't have manual mode anymore, so continuous Auto trim prob isn't doin anything for me. Other than confirming my servo centerness.

Yes, as far as I understand it.
If you fly in a straight line and switch from ACRO to MANUAL, it keeps flying in a straight line.
It's enabled/disabled with a slider in the Configurator.
In your case, just make sure it's switched off.

**Comment by P-I-Engineer** [comment]
Thanks, I'll use just for servo centering check on my maidens from now on.

**Comment by MartinHugh** [comment]
> Auto trim is for Acro

Corrections welcome, but I understood that Auto trim is calculated in all modes except Manual.

**Comment by JulioCesarMatias** [comment]
> > > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> > 
> > 
> > but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.
> 
> I understand. Problem is that I seldom run into the config wipe issue. And the last time it was because we unplugged a lipo just after disarm. So I'm not able to test and give a conclusive answer. Sorry.
> 
> A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates. They run a MATEKF405SE and MATEKF405TE_SD

I'll leave the .hex files here in case they are interested in testing

[inav_6.0.0_MATEKF405TE_SD.zip](https://github.com/iNavFlight/inav/files/9128426/inav_6.0.0_MATEKF405TE_SD.zip)
[inav_6.0.0_MATEKF405SE.zip](https://github.com/iNavFlight/inav/files/9128427/inav_6.0.0_MATEKF405SE.zip)

**Comment by MrD-RC** [comment]
> Is continuous Auto trim only used for flying in manual? I don't have manual mode anymore, so continuous Auto trim prob isn't doin anything for me. Other than confirming my servo centerness.

No, autotrim is used in all flight modes. However, it is only adjusted in stabilised flight modes.

**Comment by P-I-Engineer** [comment]
it's purpose is only for giving you "trimmed" servos in manual though,
correct?

**Comment by MrD-RC** [comment]
No. Trimmed servos help in all flight modes. Why should the stabilisation have to compensate for a badly trimmed plane? Likewise, stabilisation should be an excuse not to trim.

**Comment by P-I-Engineer** [comment]
here's what i'm thinking.

you fly your maiden with auto servo trim on, you land, you look at your center values, if they're way off you mechanically adjust them.

fly again to recheck.

after that if you don't fly in manual, why would your mixer care about keeping track of the trim position? the mixer should just keep changing the output till it reaches desired response.

**Comment by MrD-RC** [comment]
Part 1 (until _fly again and recheck_): 100% That is what should be done. This will give a good mechanical baseline. 

Part 2: Why should the might mode come in to it? If the conditions have changed so that it is out of trim, the trim should be fine tuned. Why should the flight controller have to work to constantly correct a behaviour when it can be trimmed out? Then all it needs to worry about the actual stabilisation tasks.

**Comment by MartinHugh** [comment]
Just to report this happened to me this morning.  
Yesterday I was doing some arming checks and left this aircraft for 10  minutes powered up to collects sats.
After this was done I removed the battery as normal.

This morning after a 10 mile drive to the flying field I discover the FC has been wiped.

Are we sure we understand all the possible causes of this problem?

(Target MATEK405SE)

**Comment by MartinHugh** [comment]
Same aircraft, iNav 5.  I removed the battery after it had been sat disarmed for about 5 minutes.  On the next power up, the config had been wiped.

**Comment by JulioCesarMatias** [comment]
> Mesma aeronave, iNav 5. Eu removi a bateria depois que ela ficou desarmada por cerca de 5 minutos. Na próxima inicialização, a configuração foi apagada.

In my comment above, there is a build for MatekF405SE, could you test it?

**Comment by JulioCesarMatias** [comment]
> > > > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> > > 
> > > 
> > > but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.
> > 
> > 
> > I understand. Problem is that I seldom run into the config wipe issue. And the last time it was because we unplugged a lipo just after disarm. So I'm not able to test and give a conclusive answer. Sorry.
> > A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates. They run a MATEKF405SE and MATEKF405TE_SD
> 
> I'll leave the .hex files here in case they are interested in testing
> 
> [inav_6.0.0_MATEKF405TE_SD.zip](https://github.com/iNavFlight/inav/files/9128426/inav_6.0.0_MATEKF405TE_SD.zip) [inav_6.0.0_MATEKF405SE.zip](https://github.com/iNavFlight/inav/files/9128427/inav_6.0.0_MATEKF405SE.zip)

@MartinHugh

**Comment by MartinHugh** [comment]
> > Same aircraft, iNav 5. I removed the battery after it had been sat disarmed for about 5 minutes. On the next power up, the config had been wiped.
> 
> In my comment above, there is a build for MatekF405SE, could you test it?

@JulioCesarMatias Thanks,  I first need to finish setting up the tune with the standard build.  I have having some issues dialling in the mechanical trim at the moment.  The biggest problem I see is that it is difficult to deduce anything certain when testing for intermittent faults.  Nevertheless, once I get passed the setup and tune of this particular aircraft I will test your firmware.  Thanks,  Can you again summarise the changes and how this affects what I need to do (if anything)

**Comment by JulioCesarMatias** [comment]
> > > Same aircraft, iNav 5. I removed the battery after it had been sat disarmed for about 5 minutes. On the next power up, the config had been wiped.
> > 
> > 
> > In my comment above, there is a build for MatekF405SE, could you test it?
> 
> @JulioCesarMatias Thanks, I first need to finish setting up the tune with the standard build. I have having some issues dialling in the mechanical trim at the moment. The biggest problem I see is that it is difficult to deduce anything certain when testing for intermittent faults. Nevertheless, once I get passed the setup and tune of this particular aircraft I will test your firmware. Thanks, Can you again summarise the changes and how this affects what I need to do (if anything)

The change I made is simple, I described the changes in a comment above. And nothing additional is necessary, just record in your FC and test

**Comment by JulioCesarMatias** [comment]
> > > Same aircraft, iNav 5. I removed the battery after it had been sat disarmed for about 5 minutes. On the next power up, the config had been wiped.
> > 
> > 
> > In my comment above, there is a build for MatekF405SE, could you test it?
> 
> @JulioCesarMatias Thanks, I first need to finish setting up the tune with the standard build. I have having some issues dialling in the mechanical trim at the moment. The biggest problem I see is that it is difficult to deduce anything certain when testing for intermittent faults. Nevertheless, once I get passed the setup and tune of this particular aircraft I will test your firmware. Thanks, Can you again summarise the changes and how this affects what I need to do (if anything)

And you can finish your tweaks with my compilation, it won't make a difference

**Comment by MartinHugh** [comment]
> > > > Same aircraft, iNav 5. I removed the battery after it had been sat disarmed for about 5 minutes. On the next power up, the config had been wiped.
> > > 
> > > 
> > > In my comment above, there is a build for MatekF405SE, could you test it?
> > 
> > 
> > @JulioCesarMatias Thanks, I first need to finish setting up the tune with the standard build. I have having some issues dialling in the mechanical trim at the moment. The biggest problem I see is that it is difficult to deduce anything certain when testing for intermittent faults. Nevertheless, once I get passed the setup and tune of this particular aircraft I will test your firmware. Thanks, Can you again summarise the changes and how this affects what I need to do (if anything)
> 
> And you can finish your settings with my build, it won't make a difference

Understood. I will finish the setup with the released version first and then switch.  Note it is windy here for the next 3 days.

**Comment by cpwilson** [comment]
> > > > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> > > 
> > > 
> > > but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.
> > 
> > 
> > I understand. Problem is that I seldom run into the config wipe issue. And the last time it was because we unplugged a lipo just after disarm. So I'm not able to test and give a conclusive answer. Sorry.
> > A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates. They run a MATEKF405SE and MATEKF405TE_SD
> 
> I'll leave the .hex files here in case they are interested in testing
> 
> [inav_6.0.0_MATEKF405TE_SD.zip](https://github.com/iNavFlight/inav/files/9128426/inav_6.0.0_MATEKF405TE_SD.zip) [inav_6.0.0_MATEKF405SE.zip](https://github.com/iNavFlight/inav/files/9128427/inav_6.0.0_MATEKF405SE.zip)

Will give this a test and report back.

**Comment by cpwilson** [comment]
> > > > > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> > > > 
> > > > 
> > > > but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.
> > > 
> > > 
> > > I understand. Problem is that I seldom run into the config wipe issue. And the last time it was because we unplugged a lipo just after disarm. So I'm not able to test and give a conclusive answer. Sorry.
> > > A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates. They run a MATEKF405SE and MATEKF405TE_SD
> > 
> > 
> > I'll leave the .hex files here in case they are interested in testing
> > [inav_6.0.0_MATEKF405TE_SD.zip](https://github.com/iNavFlight/inav/files/9128426/inav_6.0.0_MATEKF405TE_SD.zip) [inav_6.0.0_MATEKF405SE.zip](https://github.com/iNavFlight/inav/files/9128427/inav_6.0.0_MATEKF405SE.zip)
> 
> Will give this a test and report back.

OK tested 6.0.0 this morning and still have the config being wiped.

Steps taken
I took a full dump of the config from 5.0.0
Flashed 6.0.0 to the FC
Restored the full dump as using configurator for 5.0 unable to apply preset as only have cli access
Checked config applied and all was working
Power cycled the FC a few times, 4 or 5 and no issues FC seemed to boot up fine
Powered off and left for 40 mins and when powered on FC config is wipe

Note, power cycles were done with flight battery



Have attached full dump file from 5.0.0 and diff from 6.0.0 after config saved and also one after config wipe

[6.0.0 diff all after config wiped.txt](https://github.com/iNavFlight/inav/files/9173572/6.0.0.diff.all.after.config.wiped.txt)
[6.0.0 diff all after config.txt](https://github.com/iNavFlight/inav/files/9173573/6.0.0.diff.all.after.config.txt)
[5.0.0 full dump.txt](https://github.com/iNavFlight/inav/files/9173574/5.0.0.full.dump.txt)
d

**Comment by MartinHugh** [comment]
> > > > > Same aircraft, iNav 5. I removed the battery after it had been sat disarmed for about 5 minutes. On the next power up, the config had been wiped.
> > > > 
> > > > 
> > > > In my comment above, there is a build for MatekF405SE, could you test it?
> > > 
> > > 
> > > @JulioCesarMatias Thanks, I first need to finish setting up the tune with the standard build. I have having some issues dialling in the mechanical trim at the moment. The biggest problem I see is that it is difficult to deduce anything certain when testing for intermittent faults. Nevertheless, once I get passed the setup and tune of this particular aircraft I will test your firmware. Thanks, Can you again summarise the changes and how this affects what I need to do (if anything)
> > 
> > 
> > And you can finish your settings with my build, it won't make a difference
> 
> Understood. I will finish the setup with the released version first and then switch. Note it is windy here for the next 3 days.

OK I have the mechanical trims set up and iNav tuned on 5.0.0.  I am in a position to test a bug fix version now.  I will will hang fire though until the results of @cpwilson have been reviewed.  As I expect another test on this exact build will have little value now.

**Comment by MartinHugh** [comment]
> > > > > > Ahhh, that makes sense. That means your solution is the second best thing v.s. not saving to EEPROM at all. Definitely recommended for v5.1.0 if you ask me.
> > > > > 
> > > > > 
> > > > > but I need to make sure that this will alleviate the problems... so I need several people who have the auto-erase problem to test my changes. I don't have and never had the auto-erase problem.
> > > > 
> > > > 
> > > > I understand. Problem is that I seldom run into the config wipe issue. And the last time it was because we unplugged a lipo just after disarm. So I'm not able to test and give a conclusive answer. Sorry.
> > > > A bit upwards in this thread @cpwilson and/or @dcamptell might be good candidates. They run a MATEKF405SE and MATEKF405TE_SD
> > > 
> > > 
> > > I'll leave the .hex files here in case they are interested in testing
> > > [inav_6.0.0_MATEKF405TE_SD.zip](https://github.com/iNavFlight/inav/files/9128426/inav_6.0.0_MATEKF405TE_SD.zip) [inav_6.0.0_MATEKF405SE.zip](https://github.com/iNavFlight/inav/files/9128427/inav_6.0.0_MATEKF405SE.zip)
> > 
> > 
> > Will give this a test and report back.
> 
> OK tested 6.0.0 this morning and still have the config being wiped.
> 
> Steps taken I took a full dump of the config from 5.0.0 Flashed 6.0.0 to the FC Restored the full dump as using configurator for 5.0 unable to apply preset as only have cli access Checked config applied and all was working Power cycled the FC a few times, 4 or 5 and armed board whilst powered on and no issues FC seemed to boot up fine Powered off and left for 40 mins and when powered on FC config is wiped
> 
> Note, power cycles were done with flight battery
> 
> Have attached full dump file from 5.0.0 and diff from 6.0.0 after config saved and also one after config wipe
> 
> [6.0.0 diff all after config wiped.txt](https://github.com/iNavFlight/inav/files/9173572/6.0.0.diff.all.after.config.wiped.txt) [6.0.0 diff all after config.txt](https://github.com/iNavFlight/inav/files/9173573/6.0.0.diff.all.after.config.txt) [5.0.0 full dump.txt](https://github.com/iNavFlight/inav/files/9173574/5.0.0.full.dump.txt)

Thank you for doing this @cpwilson.   Can you confirm you were testing short power on/power off cycles (which normally attempt to write to the EEPROM on boot up) ?

**Comment by cpwilson** [comment]
No problems at all.    I simulated 'real life' usage as if I was at the field with powering on, arming the FC then disarming it an powering off, all via the flight pack.    Did that a few times and had a couple of USB connections to it as well to check config still there.
It wasn't until it was last powered off and left for 40 mins that when powered on I had a wiped config.

So in answer to the short power on/power off cycles question I would say my shortest power cycle was possibly 2-3 mins.

**Comment by MartinHugh** [comment]
> No problems at all. I simulated 'real life' usage as if I was at the field with powering on, arming the FC then disarming it an powering off, all via the flight pack. Did that a few times and had a couple of USB connections to it as well to check config still there. It wasn't until it was last powered off and left for 40 mins that when powered on I had a wiped config.
> 
> So in answer to the short power on/power off cycles question I would say my shortest power cycle was possibly 2-3 mins.

@cpwilson, just a couple of other things to put the results in context if that is OK?

Did you arm the aircraft
Did you fly the aircraft
Do you have autolevel permanently enabled in the configuration.

Many thanks.

**Comment by tipoman9** [comment]
After upgrading from 3.0.2 to 5.0.0 this same issue started to happen to me also. Really annoying.
Board is MatekF405SE. 
My other INAV 5.0.0 with OmnibusF4Pro does not seem to have this problem...

**Comment by cpwilson** [comment]
> > No problems at all. I simulated 'real life' usage as if I was at the field with powering on, arming the FC then disarming it an powering off, all via the flight pack. Did that a few times and had a couple of USB connections to it as well to check config still there. It wasn't until it was last powered off and left for 40 mins that when powered on I had a wiped config.
> > So in answer to the short power on/power off cycles question I would say my shortest power cycle was possibly 2-3 mins.
> 
> @cpwilson, just a couple of other things to put the results in context if that is OK?
> 
> Did you arm the aircraft Did you fly the aircraft Do you have autolevel permanently enabled in the configuration.
> 
> Many thanks.

Did you arm the aircraft
Yes, FC was armed with enough sats etc as it I was to fly it

Did you fly the aircraft
No, only bench tested it, but did cycle through my modes, acro, horizon etc to check they all worked

Do you have autolevel permanently enabled in the configuration.
If you mean servo autotrim, then yes it's always enabled, 'feature -FW_AUTOTRIM'

**Comment by 0crap** [comment]
Can you try without servo autotrim? (Recommended for everyone running into this issue.)
Autotrim can write to EEPROM and maybe related to the issue.
Since you did not fly the craft I don't think it's your root cause, but give it a try anyway.
Thx.

**Comment by MartinHugh** [comment]
> > > No problems at all. I simulated 'real life' usage as if I was at the field with powering on, arming the FC then disarming it an powering off, all via the flight pack. Did that a few times and had a couple of USB connections to it as well to check config still there. It wasn't until it was last powered off and left for 40 mins that when powered on I had a wiped config.
> > > So in answer to the short power on/power off cycles question I would say my shortest power cycle was possibly 2-3 mins.
> > 
> > 
> > @cpwilson, just a couple of other things to put the results in context if that is OK?
> > Did you arm the aircraft Did you fly the aircraft Do you have autolevel permanently enabled in the configuration.
> > Many thanks.
> 
> Did you arm the aircraft Yes, FC was armed with enough sats etc as it I was to fly it
> 
> Did you fly the aircraft No, only bench tested it, but did cycle through my modes, acro, horizon etc to check they all worked
> 
> Do you have autolevel permanently enabled in the configuration. If you mean servo autotrim, then yes it's always enabled, 'feature -FW_AUTOTRIM'

@cpwilson thank you for the info, and sorry, yes I meant autotrim.

**Comment by MartinHugh** [comment]
> Can you try without servo autotrim? (Recommended for everyone running into this issue.) Autotrim can write to EEPROM and maybe related to the issue. Since you did not fly the craft I don't think it's your root cause, but give it a try anyway. Thx.

Can anyone clarify *when* autotrim is written, I thought is was only written to EEPROM on disarm.

**Comment by 0crap** [comment]
> Can anyone clarify _when_ autotrim is written, I thought is was only written to EEPROM on disarm.

Correct.

**Comment by viplexa** [comment]
Hello everyone, in general, I can repeat this problem forever, I attach a video, I don't save anything, I just connect and disconnect the usb and the configurator several times and after a few attempts the settings are cleared.

Matek F405S
INAV 5.0.0

https://user-images.githubusercontent.com/11394073/180667594-72a095b9-81f4-4ef5-9511-cd6e76c6c2a3.mp4

**Comment by cpwilson** [comment]
> > Can you try without servo autotrim? (Recommended for everyone running into this issue.) Autotrim can write to EEPROM and maybe related to the issue. Since you did not fly the craft I don't think it's your root cause, but give it a try anyway. Thx.
> 
> Can anyone clarify _when_ autotrim is written, I thought is was only written to EEPROM on disarm.

OK so with AUTOTRIM disabled I managed about 10 power cycles with arming the FC with no issues.
Left for 2hrs and it was still OK :) with several more arm/disarm cycles.
Left overnight and plugged up this morning and config gone :(

**Comment by MartinHugh** [comment]
> > > Can you try without servo autotrim? (Recommended for everyone running into this issue.) Autotrim can write to EEPROM and maybe related to the issue. Since you did not fly the craft I don't think it's your root cause, but give it a try anyway. Thx.
> > 
> > 
> > Can anyone clarify _when_ autotrim is written, I thought is was only written to EEPROM on disarm.
> 
> OK so with AUTOTRIM disabled I managed about 10 power cycles with arming the FC with no issues. Left for 2hrs and it was still OK :) with several more arm/disarm cycles. Left overnight and plugged up this morning and config gone :(

So even if you did pull the battery immediately after disarming, we know it was not down to the autotrim save.

The next question is, does anything else get written to EEPROM when the aircraft is disarmed ?

**Comment by 0crap** [comment]
On v5.0.0 yes, that will be fixed in v5.1.0 with PR #8216

**Comment by 0crap** [comment]
> For that reason alone, I think autotrim should only be written when the pilot saves explicitly using the sticks.

Agreed.

**Comment by tipoman9** [comment]
I've discovered the cause. You need to unplug and plug quickly and it happens. 
I'm 100% sure. I've plugged in the plane today, checked my OSD and saw "no RC link" message - I had forgotten to turn on the TX. Since ELRS needs the TX running when the RX is turned on, my only option was to quickly reset the latter, which requires a quick disconnect-reconnect. And after that my customized OSD changed to the default one, no serve response, etc.

**Comment by ishay320** [comment]
I had it 2 times on wingFC on 5.0.0, trying now on 5.1.0

**Comment by ahmadhassan11** [comment]
I use matek h743 v2 board for Diaton r690 aircraft
    Yes you were right, I forgot to save
thank you
have a nice day

**Comment by P-I-Engineer** [comment]
Calculated, but actually used only in manual.

**Comment by b14ckyy** [comment]
@P-I-Engineer Nope it trims the servos on Mixer level so the trim is actually used Everywhere. 
Trimmed servos release the Integral value of the PID controller to have less work to do, when rate setpoints are 0.

**Comment by krasen1234** [comment]
Still not fixed, just happened on a Speedybee 405 mini and inav 7.1.2

**Comment by b14ckyy** [comment]
@krasen1234 
This ticket is over 2 years closed now. 
Please open a proper new issue ticket with all the necessary info and steps you did when it happened.

**Comment by sensei-hacker** [comment]
> This just happened to me! Matek F405 TE mini - unplugged power then powered on again quickly and it reset the FC to factory settings iNav 8.0

Please see the comment just above yours.

Also you might read any of the other comments, where it's explained that if you yank the battery power while the settings are being written, then yep you'll have corrupted settings. Half old settings, half new, which is invalid so the settings are cleared.
